File: servico-clientes/servidor.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import uuid
import threading
import time
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime
import json
import mysql.connector
from mysql.connector import Error

app = Flask(__name__)
CORS(app)

# Configuração
PORTA = 3002

# Configuração do banco de dados
DB_CONFIG = {
    'host': 'mysql',
    'port': 3306,
    'user': 'vendas_user',
    'password': 'vendas123',
    'database': 'vendas_db',
    'charset': 'utf8mb4'
}

def get_db_connection():
    """Criar conexão com o banco de dados"""
    try:
        connection = mysql.connector.connect(**DB_CONFIG)
        return connection
    except Error as e:
        print(f"Erro ao conectar com MySQL: {e}")
        return None

# Armazenamento em memória (simulando banco de dados)
clientes = [
    {
        'id': str(uuid.uuid4()),
        'nome': 'João Silva',
        'telefone': '(11) 99999-9999',
        'data_criacao': datetime.now().isoformat()
    },
    {
        'id': str(uuid.uuid4()),
        'nome': 'Maria Santos',
        'telefone': '(11) 88888-8888',
        'data_criacao': datetime.now().isoformat()
    }
]

# Lock para thread safety
clientes_lock = threading.Lock()

# Função para simular processamento pesado com multithreading
def processar_dados_clientes_paralelo(clientes_list):
    """
    Simula processamento pesado usando multithreading
    """
    def calcular_estatisticas():
        """Calcula estatísticas básicas dos clientes"""
        time.sleep(0.1)  # Simula processamento
        return {
            'total_clientes': len(clientes_list),
            'timestamp': datetime.now().isoformat()
        }
    
    def validar_telefones():
        """Valida formato dos telefones"""
        time.sleep(0.1)  # Simula processamento
        telefones_validos = 0
        for cliente in clientes_list:
            if len(cliente['telefone']) >= 10:
                telefones_validos += 1
        return {'telefones_validos': telefones_validos}
    
    def gerar_relatorio():
        """Gera relatório de clientes"""
        time.sleep(0.1)  # Simula processamento
        return {
            'relatorio': f'Relatório gerado com {len(clientes_list)} clientes',
            'data_geracao': datetime.now().isoformat()
        }
    
    # Executar tarefas em paralelo usando ThreadPoolExecutor
    with ThreadPoolExecutor(max_workers=3) as executor:
        # Submeter tarefas para execução paralela
        futuro_estatisticas = executor.submit(calcular_estatisticas)
        futuro_validacao = executor.submit(validar_telefones)
        futuro_relatorio = executor.submit(gerar_relatorio)
        
        # Aguardar resultados
        estatisticas = futuro_estatisticas.result()
        validacao = futuro_validacao.result()
        relatorio = futuro_relatorio.result()
    
    return {
        'estatisticas': estatisticas,
        'validacao': validacao,
        'relatorio': relatorio
    }

# Rotas

@app.route('/clientes', methods=['GET'])
def listar_clientes():
    """Listar todos os clientes com processamento paralelo"""
    try:
        # Buscar clientes do banco de dados
        connection = get_db_connection()
        if not connection:
            return jsonify({'erro': 'Erro ao conectar com o banco de dados'}), 500
        
        cursor = connection.cursor(dictionary=True)
        cursor.execute("SELECT * FROM clientes ORDER BY data_criacao DESC")
        clientes_db = cursor.fetchall()
        
        # Converter valores para o formato esperado
        clientes_formatados = []
        for cliente in clientes_db:
            clientes_formatados.append({
                'id': cliente['id'],
                'nome': cliente['nome'],
                'telefone': cliente['telefone'],
                'data_criacao': cliente['data_criacao'].isoformat() if cliente['data_criacao'] else None
            })
        
        cursor.close()
        connection.close()
        
        # Usar multithreading para processar dados
        dados_processados = processar_dados_clientes_paralelo(clientes_formatados)
        
        return jsonify({
            'clientes': clientes_formatados,
            'dados_processados': dados_processados,
            'total': len(clientes_formatados)
        })
    except Exception as e:
        return jsonify({'erro': f'Erro interno do servidor: {str(e)}'}), 500

@app.route('/clientes/<cliente_id>', methods=['GET'])
def buscar_cliente(cliente_id):
    """Buscar cliente por ID"""
    try:
        # Buscar cliente no banco de dados
        connection = get_db_connection()
        if not connection:
            return jsonify({'erro': 'Erro ao conectar com o banco de dados'}), 500
        
        cursor = connection.cursor(dictionary=True)
        cursor.execute("SELECT * FROM clientes WHERE id = %s", (cliente_id,))
        cliente_db = cursor.fetchone()
        
        cursor.close()
        connection.close()
        
        if not cliente_db:
            return jsonify({'erro': 'Cliente não encontrado'}), 404
        
        # Converter para o formato esperado
        cliente = {
            'id': cliente_db['id'],
            'nome': cliente_db['nome'],
            'telefone': cliente_db['telefone'],
            'data_criacao': cliente_db['data_criacao'].isoformat() if cliente_db['data_criacao'] else None
        }
        
        return jsonify(cliente)
        
    except Exception as e:
        return jsonify({'erro': f'Erro interno do servidor: {str(e)}'}), 500

@app.route('/clientes', methods=['POST'])
def cadastrar_cliente():
    """Cadastrar novo cliente"""
    dados = request.get_json()
    
    # Validação básica
    if not dados or 'nome' not in dados or 'telefone' not in dados:
        return jsonify({'erro': 'Nome e telefone são obrigatórios'}), 400
    
    nome = dados['nome'].strip()
    telefone = dados['telefone'].strip()
    
    if not nome or not telefone:
        return jsonify({'erro': 'Nome e telefone não podem estar vazios'}), 400
    
    try:
        # Conectar com o banco de dados
        connection = get_db_connection()
        if not connection:
            return jsonify({'erro': 'Erro ao conectar com o banco de dados'}), 500
        
        cursor = connection.cursor()
        cliente_id = str(uuid.uuid4())
        
        # Inserir cliente no banco
        cursor.execute(
            "INSERT INTO clientes (id, nome, telefone) VALUES (%s, %s, %s)",
            (cliente_id, nome, telefone)
        )
        
        connection.commit()
        cursor.close()
        connection.close()
        
        # Retornar cliente criado
        novo_cliente = {
            'id': cliente_id,
            'nome': nome,
            'telefone': telefone,
            'data_criacao': datetime.now().isoformat()
        }
        
        return jsonify(novo_cliente), 201
        
    except Exception as e:
        return jsonify({'erro': f'Erro ao salvar cliente: {str(e)}'}), 500

@app.route('/clientes/<cliente_id>', methods=['PUT'])
def atualizar_cliente(cliente_id):
    """Atualizar cliente existente"""
    dados = request.get_json()
    
    if not dados or 'nome' not in dados or 'telefone' not in dados:
        return jsonify({'erro': 'Nome e telefone são obrigatórios'}), 400
    
    nome = dados['nome'].strip()
    telefone = dados['telefone'].strip()
    
    if not nome or not telefone:
        return jsonify({'erro': 'Nome e telefone não podem estar vazios'}), 400
    
    try:
        # Conectar com o banco de dados
        connection = get_db_connection()
        if not connection:
            return jsonify({'erro': 'Erro ao conectar com o banco de dados'}), 500
        
        cursor = connection.cursor()
        
        # Verificar se o cliente existe
        cursor.execute("SELECT * FROM clientes WHERE id = %s", (cliente_id,))
        cliente_existente = cursor.fetchone()
        
        if not cliente_existente:
            cursor.close()
            connection.close()
            return jsonify({'erro': 'Cliente não encontrado'}), 404
        
        # Atualizar cliente no banco
        cursor.execute(
            "UPDATE clientes SET nome = %s, telefone = %s WHERE id = %s",
            (nome, telefone, cliente_id)
        )
        
        connection.commit()
        cursor.close()
        connection.close()
        
        # Retornar cliente atualizado
        cliente_atualizado = {
            'id': cliente_id,
            'nome': nome,
            'telefone': telefone,
            'data_criacao': cliente_existente[3].isoformat() if cliente_existente[3] else None,
            'data_atualizacao': datetime.now().isoformat()
        }
        
        return jsonify(cliente_atualizado)
        
    except Exception as e:
        return jsonify({'erro': f'Erro ao atualizar cliente: {str(e)}'}), 500

@app.route('/clientes/<cliente_id>', methods=['DELETE'])
def deletar_cliente(cliente_id):
    """Deletar cliente"""
    try:
        # Conectar com o banco de dados
        connection = get_db_connection()
        if not connection:
            return jsonify({'erro': 'Erro ao conectar com o banco de dados'}), 500
        
        cursor = connection.cursor()
        
        # Verificar se o cliente existe
        cursor.execute("SELECT * FROM clientes WHERE id = %s", (cliente_id,))
        cliente_existente = cursor.fetchone()
        
        if not cliente_existente:
            cursor.close()
            connection.close()
            return jsonify({'erro': 'Cliente não encontrado'}), 404
        
        # Deletar cliente do banco
        cursor.execute("DELETE FROM clientes WHERE id = %s", (cliente_id,))
        
        connection.commit()
        cursor.close()
        connection.close()
        
        return jsonify({'mensagem': 'Cliente removido com sucesso'})
        
    except Exception as e:
        return jsonify({'erro': f'Erro ao deletar cliente: {str(e)}'}), 500

@app.route('/saude', methods=['GET'])
def health_check():
    """Health check do serviço"""
    return jsonify({
        'status': 'OK',
        'servico': 'servico-clientes',
        'timestamp': datetime.now().isoformat(),
        'threads_ativas': threading.active_count()
    })

@app.route('/estatisticas', methods=['GET'])
def estatisticas_clientes():
    """Endpoint específico para estatísticas usando multithreading"""
    try:
        dados_processados = processar_dados_clientes_paralelo()
        return jsonify(dados_processados)
    except Exception as e:
        return jsonify({'erro': f'Erro ao processar estatísticas: {str(e)}'}), 500

if __name__ == '__main__':
    print(f"🚀 Serviço de Clientes rodando na porta {PORTA}")
    print(f"📊 Health check: http://localhost:{PORTA}/saude")
    print(f"👥 API Clientes: http://localhost:{PORTA}/clientes")
    print(f"📈 Estatísticas: http://localhost:{PORTA}/estatisticas")
    
    app.run(host='0.0.0.0', port=PORTA, debug=True, threaded=True)

File: servico-produtos/servidor.js
const express = require('express');
const cors = require('cors');
const { v4: uuidv4 } = require('uuid');
const { Worker } = require('worker_threads');
const path = require('path');
const mysql = require('mysql2/promise');

const app = express();
const PORTA = process.env.PORTA || 3001;

// Configuração do banco de dados
const dbConfig = {
  host: process.env.DB_HOST || 'localhost',
  port: process.env.DB_PORT || 3306,
  user: process.env.DB_USER || 'vendas_user',
  password: process.env.DB_PASSWORD || 'vendas123',
  database: process.env.DB_NAME || 'vendas_db',
  waitForConnections: true,
  connectionLimit: 10,
  queueLimit: 0
};

// Pool de conexões MySQL
const pool = mysql.createPool(dbConfig);

// Middleware
app.use(cors());
app.use(express.json());

// Worker thread para processamento paralelo de cálculos
function calcularEstatisticasProdutos(produtos) {
  return new Promise((resolve, reject) => {
    const worker = new Worker(path.join(__dirname, 'trabalhadores', 'calculos.js'), {
      workerData: { produtos }
    });

    worker.on('message', (resultado) => {
      resolve(resultado);
    });

    worker.on('error', (erro) => {
      reject(erro);
    });

    worker.on('exit', (codigo) => {
      if (codigo !== 0) {
        reject(new Error(`Worker parou com código ${codigo}`));
      }
    });
  });
}

// Rotas

// GET /produtos - Listar todos os produtos
app.get('/produtos', async (req, res) => {
  try {
    // Buscar produtos do banco
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos ORDER BY data_criacao DESC'
    );
    
    // Converter valores para números
    const produtosFormatados = produtos.map(produto => ({
      ...produto,
      valor: parseFloat(produto.valor)
    }));
    
    // Usar worker thread para calcular estatísticas em paralelo
    const estatisticas = await calcularEstatisticasProdutos(produtosFormatados);
    
    res.json({
      produtos: produtosFormatados,
      estatisticas,
      total: produtosFormatados.length
    });
  } catch (erro) {
    console.error('Erro ao listar produtos:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// GET /produtos/:id - Buscar produto por ID
app.get('/produtos/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos WHERE id = ?',
      [id]
    );
    
    if (produtos.length === 0) {
      return res.status(404).json({ erro: 'Produto não encontrado' });
    }
    
    // Converter valor para número
    const produto = {
      ...produtos[0],
      valor: parseFloat(produtos[0].valor)
    };
    
    res.json(produto);
  } catch (erro) {
    console.error('Erro ao buscar produto:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// POST /produtos - Cadastrar novo produto
app.post('/produtos', async (req, res) => {
  try {
    const { nome, descricao, valor, quantidade = 0, quantidade_minima = 5 } = req.body;
    
    // Validação básica
    if (!nome || !descricao || !valor) {
      return res.status(400).json({ 
        erro: 'Nome, descrição e valor são obrigatórios' 
      });
    }
    
    if (valor <= 0) {
      return res.status(400).json({ 
        erro: 'Valor deve ser maior que zero' 
      });
    }
    
    if (quantidade < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade não pode ser negativa' 
      });
    }
    
    if (quantidade_minima < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade mínima não pode ser negativa' 
      });
    }
    
    const id = uuidv4();
    const [result] = await pool.execute(
      'INSERT INTO produtos (id, nome, descricao, valor, quantidade, quantidade_minima) VALUES (?, ?, ?, ?, ?, ?)',
      [id, nome.trim(), descricao.trim(), parseFloat(valor), parseInt(quantidade), parseInt(quantidade_minima)]
    );
    
    // Buscar o produto criado
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos WHERE id = ?',
      [id]
    );
    
    res.status(201).json(produtos[0]);
  } catch (erro) {
    console.error('Erro ao criar produto:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// PUT /produtos/:id - Atualizar produto
app.put('/produtos/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { nome, descricao, valor, quantidade, quantidade_minima } = req.body;
    
    // Validação básica
    if (!nome || !descricao || !valor) {
      return res.status(400).json({ 
        erro: 'Nome, descrição e valor são obrigatórios' 
      });
    }
    
    if (valor <= 0) {
      return res.status(400).json({ 
        erro: 'Valor deve ser maior que zero' 
      });
    }
    
    if (quantidade !== undefined && quantidade < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade não pode ser negativa' 
      });
    }
    
    if (quantidade_minima !== undefined && quantidade_minima < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade mínima não pode ser negativa' 
      });
    }
    
    const [result] = await pool.execute(
      'UPDATE produtos SET nome = ?, descricao = ?, valor = ?, quantidade = ?, quantidade_minima = ? WHERE id = ?',
      [nome.trim(), descricao.trim(), parseFloat(valor), parseInt(quantidade || 0), parseInt(quantidade_minima || 5), id]
    );
    
    if (result.affectedRows === 0) {
      return res.status(404).json({ erro: 'Produto não encontrado' });
    }
    
    // Buscar o produto atualizado
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos WHERE id = ?',
      [id]
    );
    
    res.json(produtos[0]);
  } catch (erro) {
    console.error('Erro ao atualizar produto:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// PUT /produtos/:id/estoque - Atualizar apenas o estoque
app.put('/produtos/:id/estoque', async (req, res) => {
  try {
    const { id } = req.params;
    const { quantidade, quantidade_minima } = req.body;
    
    if (quantidade === undefined && quantidade_minima === undefined) {
      return res.status(400).json({ 
        erro: 'Pelo menos um campo de estoque deve ser fornecido' 
      });
    }
    
    if (quantidade !== undefined && quantidade < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade não pode ser negativa' 
      });
    }
    
    if (quantidade_minima !== undefined && quantidade_minima < 0) {
      return res.status(400).json({ 
        erro: 'Quantidade mínima não pode ser negativa' 
      });
    }
    
    let query = 'UPDATE produtos SET ';
    let params = [];
    
    if (quantidade !== undefined) {
      query += 'quantidade = ?';
      params.push(parseInt(quantidade));
    }
    
    if (quantidade_minima !== undefined) {
      if (quantidade !== undefined) {
        query += ', ';
      }
      query += 'quantidade_minima = ?';
      params.push(parseInt(quantidade_minima));
    }
    
    query += ' WHERE id = ?';
    params.push(id);
    
    const [result] = await pool.execute(query, params);
    
    if (result.affectedRows === 0) {
      return res.status(404).json({ erro: 'Produto não encontrado' });
    }
    
    // Buscar o produto atualizado
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos WHERE id = ?',
      [id]
    );
    
    res.json(produtos[0]);
  } catch (erro) {
    console.error('Erro ao atualizar estoque:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// GET /produtos/estoque-baixo - Listar produtos com estoque baixo
app.get('/produtos/estoque-baixo', async (req, res) => {
  try {
    const [produtos] = await pool.execute(
      'SELECT * FROM produtos WHERE quantidade <= quantidade_minima ORDER BY quantidade ASC'
    );
    
    // Converter valores para números
    const produtosFormatados = produtos.map(produto => ({
      ...produto,
      valor: parseFloat(produto.valor),
      quantidade: parseInt(produto.quantidade),
      quantidade_minima: parseInt(produto.quantidade_minima)
    }));
    
    res.json({
      produtos: produtosFormatados,
      total: produtosFormatados.length
    });
  } catch (erro) {
    console.error('Erro ao listar produtos com estoque baixo:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// DELETE /produtos/:id - Deletar produto
app.delete('/produtos/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const [result] = await pool.execute(
      'DELETE FROM produtos WHERE id = ?',
      [id]
    );
    
    if (result.affectedRows === 0) {
      return res.status(404).json({ erro: 'Produto não encontrado' });
    }
    
    res.status(204).send();
  } catch (erro) {
    console.error('Erro ao deletar produto:', erro);
    res.status(500).json({ erro: 'Erro interno do servidor' });
  }
});

// Health check
app.get('/saude', (req, res) => {
  res.json({ 
    status: 'OK', 
    servico: 'servico-produtos',
    timestamp: new Date().toISOString()
  });
});

// Iniciar servidor
app.listen(PORTA, () => {
  console.log(`🚀 Serviço de Produtos rodando na porta ${PORTA}`);
  console.log(`📊 Health check: http://localhost:${PORTA}/saude`);
  console.log(`📦 API Produtos: http://localhost:${PORTA}/produtos`);
});

module.exports = app;
